source/braille.py

```diff
@@ -1441,11 +1441,150 @@ def _setCursor(self, info: textInfos.TextInfo):
+			if readingUnit.end > self._currentSelection.end:
+				readingUnit.end = self._currentSelection.end
+			self._withinSelection = True
+			return readingUnit
 
-	def _getSelection(self):
-		return api.getReviewPosition().copy()
+	def update(self):
+		"""Updates this region.
+		Adds functionality to properly move within selection to the next
+		and previous line and to start and end of current line, to scroll forward
+		and back to the next and previous line, and to move to focus,
+		when selection is shown.
+		"""
+		previousReadingUnit: textInfos.TextInfo | None = getattr(self, "_readingInfo", None)
+		super().update()
+		if (
+			not config.conf["braille"]["showSelection"]
+			or not self._withinSelection
+			or self._currentScriptName is None
+			or previousReadingUnit is None
+		):
+			return
+		from globalCommands import commands
+		# Using script names instead of scripts for comparison because
+		# scripts in commands are bound methods but scriptHandler.getCurrentScript
+		# returns script as function.
+		startOfNextOrPreviousLineScriptNames: set[str] = {
+			commands.script_braille_nextLine.__name__,
+			commands.script_braille_previousLine.__name__,
+			commands.script_braille_scrollForward.__name__,
+			commands.script_review_nextLine.__name__,
+			commands.script_review_previousLine.__name__,
+		}
+		startOfCurrentLineOrControlScriptNames: set[str] = {
+			commands.script_review_startOfLine.__name__,
+			commands.script_review_top.__name__,
+		}
+		endOfLineOrControlScriptNames: set[str] = {
+			commands.script_braille_scrollBack.__name__,
+			commands.script_review_endOfLine.__name__,
+			commands.script_review_bottom.__name__,
+		}
+		focusScriptNames: set[str] = {
+			commands.script_braille_toFocus.__name__,
+			commands.script_navigatorObject_toFocus.__name__,
+		}
+		log.debug(f"nykyinen {self._currentScriptName}")
+		if (
+			(
+				self._currentScriptName in startOfNextOrPreviousLineScriptNames
+				and previousReadingUnit.start != self._readingInfo.start
+				and previousReadingUnit.end != self._readingInfo.end
+			)
+			or self._currentScriptName in startOfCurrentLineOrControlScriptNames
+		):
+			# Move to start of next/current/previous line or control
+			self.brailleSelectionStart = 0
+			self.brailleSelectionEnd = 1
+		elif self._currentScriptName in endOfLineOrControlScriptNames:
+			# Scroll back to the previous line or move to end of current line or control
+			self.brailleSelectionEnd = self.rawToBraillePos[len(self.rawText) - 1]
+			self.brailleSelectionStart = self.brailleSelectionEnd - 1
+		elif self._currentScriptName in focusScriptNames:
+			# Move to start of selection
+			self.brailleSelectionEnd = self.brailleSelectionStart + 1
+		else:
+			# Do not scroll because do not know where
+			self.brailleSelectionStart = self.brailleSelectionEnd = None
+		self._currentScriptName = None
 
 	def _routeToTextInfo(self, info: textInfos.TextInfo):
+		"""Move cursor to new cell, or activate if it is already there.

```

Comment:
This change looks unrelated to the scope of reading selection when tethered to review.

Reply:
If move system caret when routing review cursor is something else than "always", second press of same routing button activates (moves caret to) current routing position. Modification of `_routeToTextInfo` tries to ensure that it happens also when "show selection" is enabled and review position is within selection. Without modification second press does not activate current routing button position (there may or likely are other alternatives to do this).